Backend/src/ml_lib/sub_models/qaqc_main_model.py

```diff
@@ -16,19 +26,116 @@ def _format_data(self, df: pd.DataFrame) -> pd.DataFrame:
+                pca2 = PCA(n_components=50, whiten=False, random_state=42)
+                reduced_embeddings_pca2 = pca2.fit_transform(embeddings)
+                tsne = TSNE(n_components=2, random_state=42, perplexity=30, n_iter=300, early_exaggeration=4, learning_rate=1000)
+                reduced_embeddings = tsne.fit_transform(reduced_embeddings_pca2)
+                point_data_df = pd.DataFrame(reduced_embeddings, columns=["x", "y"])
+
+            case "PCA":
+                pca = PCA(n_components=2, random_state=42)
+                reduced_embeddings_PCA = pca.fit_transform(embeddings)
+                point_data_df = pd.DataFrame(reduced_embeddings_PCA, columns=["x", "y"])
+
+            case "TSNE":
+                pure_tsne = TSNE(n_components=2, random_state=42, perplexity=30, n_iter=300, early_exaggeration=4, learning_rate=1000)
+                reduced_embeddings_tsne = pure_tsne.fit_transform(embeddings)
+                point_data_df = pd.DataFrame(reduced_embeddings_tsne, columns=["x", "y"])
+
+            case "UMAP":
+                umap_model = umap.UMAP(n_components=2, random_state=42, n_neighbors=15, min_dist=0.1, metric='euclidean')
+                reduced_embeddings_umap = umap_model.fit_transform(embeddings)
+                point_data_df = pd.DataFrame(reduced_embeddings_umap, columns=["x", "y"])
+
+            case _:
+                raise Exception("Invalid dimension reduction method.")
+
+        return point_data_df
+    
     def first_run(self, df: pd.DataFrame, dim: str | None) -> pd.DataFrame:
         """
         Combines the input question data with the calculated 2D point data
         """
         questions = df["text"].tolist()
         ids = [self.get_id(question) for question in questions]
-        d = {'id': ids}
-        id_df = pd.DataFrame(data=d)
-        model = SentenceTransformer("paraphrase-MiniLM-L6-v2")
-        embeddings = model.encode(questions, convert_to_tensor=True)
+        df["id"] = pd.Series(ids)
+        embeddings = self.get_embeddings(questions)
 
         point_data_df = self.dim_red(embeddings=embeddings, dim=dim)
         combined_df = pd.concat([df, point_data_df], axis=1)
-        combined_df = pd.concat([combined_df, id_df], axis=1)
-        # combined_df = combined_df.rename(columns={"coarse label": "truth"})
         return combined_df
+
+
+    def latter_run(self,df: pd.DataFrame, dim: str | None) -> pd.DataFrame:
+       
+        questions = df["text"].tolist()
+        embeddings_tensor = self.get_embeddings(questions) #This is equal to the first embeddings(1.2 in our figure), the alternative
+
+        #is to fetch these embeddings from the database
+        user_truth = df["input_label"].tolist()
+        #Calls the classifier to generate the predicted labels
+
+        try:
+            predictedLabels, new_embeddings = self.random_forest_classifier(embeddings_tensor, user_truth)
+        except Exception as e:
+            print(e)
```

Comment:
I would suggest raising e after the print. That way the correct error will be catched if another function deals with errors.

Reply:
done